docs/standard/serialization/system-text-json-migrate-from-newtonsoft-how-to.md

```diff
@@ -0,0 +1,416 @@
+
+## Converter registration precedence
+
+The `Newtonsoft.Json` registration precedence for custom converters is as follows:
+
+* Attribute on property
+* Attribute on type
+* `Converters` collection
+
+This order means that a custom converter in the `Converters` collection is overridden by a converter that is registered by applying an attribute at the type level. Both of those registrations are overridden by an attribute at the property level.
+
+The `System.Text.Json` registration precedence for custom converters is different:
+
+* Attribute on property
+* `Converters` collection
+* Attribute on type
+
+The difference here is that a custom converter in the `Converters` collection overrides an attribute at the type level. The intention behind this order of precedence is to make run-time changes override design-time choices. There's no way to change the precedence.
+
+## `Newtonsoft.Json` [JsonProperty] attribute
+
+`Newtonsoft.Json` has a `[JsonProperty]` attribute that is used for a variety of functions. `System.Text.Json` has single-purpose attributes instead of a multi-purpose attribute. The following attributes serve purposes similar to some of the `Newtonsoft.Json` attribute functions:
+
+* [[JsonPropertyName]](system-text-json-how-to.md#customize-individual-property-names)
+* [[JsonConverter]](system-text-json-converters-how-to.md#register-a-custom-converter)
+
+Other functions of the `Newtonsoft.Json` `[JsonProperty]` attribute have no equivalent attribute in `System.Text.Json`. For some of those functions, there are other ways to accomplish the same purpose, as detailed in the following sections of this article.
+
+## Character escaping
+
+During serialization, `Newtonsoft.Json` is relatively permissive about letting characters through without escaping them. (That is, it doesn't replace them with `\uxxxx` where `xxxx` is the Unicode code of the character.) `System.Text.Json` escapes more characters by default to provide defense-in-depth protections against cross-site scripting (XSS) or information disclosure attacks. `System.Text.Json` escapes all non-ASCII characters by default, so you don't need to do anything if you're using `StringEscapeHandling.EscapeNonAscii`. For information about how to override the default `System.Text.Json` behavior, see [Customize character encoding](system-text-json-how-to.md#customize-character-encoding).
+
+## Specify date format
+
+`Newtonsoft.Json` provides a `DateTimeZoneHandling` option that lets you specify the serialization format used for `DateTime` and `DateTimeOffset` properties. In `System.Text.Json`, the only format that has built-in support is ISO 8601-1:2019. To use any other format, create a custom converter. For more information, see [DateTime and DateTimeOffset support in System.Text.Json](../datetime/system-text-json-support.md).
+
+## Quoted numbers
```

Comment:
Should we mention that S.T.Json has async read/write from/to stream, potentially with workarounds for sync read/writes?

Reply:
How does that relate to Newtonsoft/ how do we differ from Newtonsoft in that respect?

Reply:
`Newtonsoft.Json` has **only** synchronous APIs to read/write from a stream via `JsonTextReader`/`JsonTextWriter`.

`System.Text.Json` has **only** asynchronous APIs to read/write from a stream via statics on `JsonSerializer`.

What happens when someone wants to read/write synchronously from/to a stream and users want to port existing code that looks like this?
https://www.newtonsoft.com/json/help/html/SerializingJSON.htm
```C#
Product product = new Product();
product.ExpiryDate = new DateTime(2008, 12, 28);

JsonSerializer serializer = new JsonSerializer();
serializer.Converters.Add(new JavaScriptDateTimeConverter());
serializer.NullValueHandling = NullValueHandling.Ignore;

using (StreamWriter sw = new StreamWriter(@"c:\json.txt"))
using (JsonWriter writer = new JsonTextWriter(sw))
{
    serializer.Serialize(writer, product);
    // {"ExpiryDate":new Date(1230375600000),"Price":0}
}
```

Reply:
How do we want to present this:  as an h3 section in the different-default-behavior  h2 section mentioning that if they use a stream they have to change to async code?  Or suggest some other approach?

Reply:
> as an h3 section in the different-default-behavior

We could do this, but since this is API difference, it doesn't fit well with the other "default behavior" headings.

I would put it as an h3 under "Scenarios using JsonSerializer that require workarounds"
Something like:
Serializing and deserializing to and from stream synchronously

The content would be something like the following:
- the built-in apis support async only and is the recommended pattern for heavy IO work, if possible
- for sync, if necessary, here are the potential workarounds (write to writer, push to stream, read data from stream and and pass to deserializer).
- support for > 2 GB payload synchronously is harder to do.

Reply:
Can you point me to code samples for each of the three workarounds? Or just a rough idea of what the code looks like and I'll implement it in the samples project.